### PR TITLE
feat(drive): composite document queries: a page plus derived sub-queries under one merged proof

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Issue being fixed or feature implemented

A feed page is one query and then, for that page, a burst of dependent ones: the quoted posts, the per-post like counts, the reposts, the authors' profiles, the viewer's own likes. Every one of them takes the page's ids or owners as input, so a client cannot issue them until the page is back, and each is its own proof against its own root. Chained queries (#4547) solved one shape of this (indexOnly inner, one join by id). This PR generalizes it into **composite document queries**: a page plus any number of sub-queries derived from its proven results, answered as ONE merged grovedb proof over one state root.

## What was done?

There is no separate composite query type: `DriveDocumentQuery` is extended with a `sub_queries` field (empty for an ordinary query, never parsed from the wire; `with_sub_queries` builds the composite form), and all composite behaviour lives on it — rs-drive core (`query/composite_document_query`), the Drive entry points (`query_composite_documents` / `_with_proof`, versioned), and the verifier (`verify_composite_documents_proof`, versioned), plus the version-table slots. The plain page-only surfaces (`query_documents(_with_flags)`, the `execute_*` wrappers, `verify_proof(_keep_serialized)`, `verify_start_at_document_in_proof`) refuse a query carrying sub-queries rather than silently proving or verifying the page alone.

`DriveChainedDocumentQuery` (#4547) is folded the same way: a chained query was always a page plus one by-id join, so it is now a `DriveDocumentQuery` whose `sub_queries` carry exactly that (built with `with_by_id_join(source_property, outer_document_type)`; `chained_join` / `validate_chained` refuse any other shape). The chained entry points, verifier, drive-abci dispatch, rich-query conversion and the proof-verifier `FromProof` bound all take `DriveDocumentQuery`; the module renames to `query/chained_document_query`, version slots unchanged.

Sub-query shapes, all bound to a source (the page, or an earlier documents sub-query) by `<field> IN <values read off the source's proven documents>`:

- **by-id join**: `bind.field == "$id"`; the source property must declare `refersTo: permanentDocument` targeting the sub-query's type, so every derived id resolves and the result is set-equal to the ids in first-appearance order (a missing referenced document is an invalid proof, exactly as chained); it takes no fixed clauses, which could legitimately miss ids;
- **documents lookup** by `$ownerId` or an indexed property, fixed clauses allowed, target may be in another contract; absence is inherent in the range proof; the limit is required on a non-unique index and forbidden when the values already bound the result (a unique index, or an indexOnly terminal with every prefix fixed). The limit caps the rows the lookup returns in total, in walk order, like an ordinary `IN` query's;
- **count** on a countable index (grouped point-lookup count keyed by the bound values);
- an unbound **sibling** documents query.

Derived values are identifiers. The page needs an explicit limit between 1 and 100 and takes no cursor or offset (paginate with a range clause); a bound field must be `$ownerId` or an identifier-typed property of the sub-query's type. A count may not select a count tree that another component descends through: grovedb cannot return the tree and walk into it in one merged selection. The shape check refuses a count and a documents component on one index path up front; the actual derived values are checked again on both entry points, so disjoint selections on the same index (the total for one post, the per-owner rows of another) stay usable.

Server: materialize the page (and any sub-query feeding a later binding), derive, `prove_query_many` over every component; root-hash bracketed and retried like chained. Every component walks in the page's direction, because the merge requires agreement: counts and by-id joins are aligned freely (their sets do not depend on direction; counts come back in key order, joins in first-appearance order), a documents sub-query the caller left unordered on its bound field inherits the page's direction, and a documents sub-query whose explicit ordering disagrees with the page is refused, since turning a limited lookup around would change the rows it returns.

Verifier: bootstrap subset pass on the page (and binding sources, assembled exactly as the server assembles them before a later binding derives from them), derive with the same builders, merge the same way, one authoritative `verify_query`, then route proved entries back to components: documents by longest base path and bound-value membership, counts by their complete terminal positions (two counts with the same bound values but different fixed clauses produce entries with identical keys, so base path plus key cannot tell them apart). Longest-base routing is only right when no documents component walks through another's base path to deeper rows, so that is checked on the concrete walks wherever the count-tree check runs. Refused: an entry no derivation asked for, a dangling join, a component whose bootstrapped documents differ from what the routed authoritative pass assigned to it, and a proven page that derives different values than the proof covers. A page-only proof (a node ignoring the sub-queries) fails closed.

Shape rules are decided from the shapes alone, so a request that validates never fails the merge for lack of data: the "no limited component at the merged root" rule is checked over the worst case (a bound sub-query that derives nothing contributes no branch, so the page and the siblings decide the root), two documents components on one index path may not carry a limit, and a merge refusal that still reaches grovedb is reported as a shape error.

Depends on two grovedb merge fixes, both merged: [dashpay/grovedb#850](https://github.com/dashpay/grovedb/pull/850) (pinned by #4597) lets a limited branch descend into a key another branch owns, which a limited page on `post` plus a by-id join on `post`, or two limited branches under one contract key once a cross-contract lookup lifts the common path, need; [dashpay/grovedb#851](https://github.com/dashpay/grovedb/pull/851) (pinned by #4602, the base of this PR) gives a synthesized split body the inputs' direction, which a descending page plus a cross-contract lookup plus a limited lookup under the page's own contract needs.

## How Has This Been Tested?

`composite_query_e2e_tests` (new `yappr-feed` fixture plus the dashpay contract for the cross-contract profile lookup):

- `should_answer_the_feed_composition_with_proof_parity`: like counts, quoted-post join (two posts quoting one target derive one id), repost lookup, page-author profiles (a missing profile is a proven absence), quoted-author profiles (bound to the join, not the page); server result equals verified result component for component
- `should_prove_the_viewers_marks_as_an_index_only_lookup`: `like` byLiker pinned on `$ownerId` with `postId IN page`
- `should_prove_an_empty_page_alone`, `should_refuse_a_page_only_proof`, `should_refuse_a_dangling_reference`, `should_tell_a_by_ids_page_from_a_join_on_the_same_type`, `should_reject_invalid_composite_shapes` (16 shapes, including a filtered join)
- `should_preserve_join_order_before_deriving_later_bindings`, `should_route_counts_by_complete_positions_including_overlapping_queries`, `should_reject_conflicting_document_directions_even_when_the_page_is_empty`, `should_reject_count_tree_descents_but_allow_disjoint_count_selections`, `should_check_count_and_document_descents_against_the_actual_bound_values`, `should_preserve_descending_documents_and_key_ordered_counts`
- `should_inherit_the_page_direction_for_unordered_lookups`: a descending by-ids page with unordered cross-contract profiles, the viewer's likes and a count merges, proves and verifies with the lookups walking descending; a limited same-contract lookup returns its capped rows from the top of the walk; the cross-contract lookup and the limited lookup together (the case grovedb #851 unblocks) do too; an explicit ascending ordering under the descending page is refused on every entry point
- `should_reject_zero_limits`, `should_reject_a_bound_field_that_is_not_identifier_typed`, `should_reject_a_limited_page_a_sub_query_below_it_could_leave_at_the_merged_root`, `should_reject_two_limited_lookups_on_one_index_path`, `should_prove_a_single_id_page_with_a_join_on_the_same_type`, plus a unit test of the walk-descent predicate
- `should_refuse_composite_queries_on_plain_surfaces`: `query_documents`, `execute_with_proof` and `verify_proof` each refuse a query carrying sub-queries

20 e2e tests and one unit test. The point-lookup count proof suite (which now shares its entry decoder with composite counts), chained, indexOnly and countable suites unchanged; clippy clean on drive with tests; verify-only feature build clean.

## Breaking Changes

None (new surface; no consensus or wire change in this PR).

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [ ] I have made corresponding changes to the documentation if needed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added composite document queries combining page results with related documents and counts.
  * Added support for generating and verifying merged proofs for composite queries.
  * Added builder support for chained by-ID joins within document queries.
  * Added composite and chained query validation with limits and binding checks.

* **Bug Fixes**
  * Plain document-query and proof APIs now reject queries containing unsupported joins or sub-queries instead of silently dropping them.

* **Tests**
  * Expanded end-to-end coverage for joins, counts, ordering, proofs, validation, and edge cases.

### CODE REVIEW

File: packages/rs-platform-version/src/version/drive_versions/drive_document_method_versions/v2.rs [hunk 1]
```diff
@@ -12,6 +12,7 @@ pub const DRIVE_DOCUMENT_METHOD_VERSIONS_V2: DriveDocumentMethodVersions =
         query: DriveDocumentQueryMethodVersions {
             query_documents: 0,
             query_chained_documents: 0,
+            query_composite_documents: 0,
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-platform-version/src/version/drive_versions/drive_document_method_versions/v2.rs:15]: Kept as is: `DriveDocumentQueryMethodVersions` is a plain struct, so a new method needs its slot in every table literal (V1 through V4) or the crate does not compile; this is how `query_chained_documents` landed in #4547 too. Every existing table gets `0`, the only version of the new method, so protocol versions 10 and 11 dispatch exactly as before, and nothing serializes or hashes these tables. Marking the slot as unchanged behaviour is the existing-table edit the module contract allows.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/rs-drive/src/query/mod.rs [hunk 2]
```diff
@@ -1089,6 +1102,25 @@ pub struct DriveDocumentQuery<'a> {
...
+    /// `query_composite_documents_with_proof` and verified by
+    /// `verify_composite_documents_proof`, and the plain
+    /// query/proof/verify surfaces refuse it rather than silently prove
+    /// the page alone.
+    ///
+    /// Never parsed from the wire: every `from_cbor` / `from_value` /
+    /// `from_typed_clauses` entry point leaves this empty; composite
+    /// requests are built programmatically (see
+    /// [`Self::with_sub_queries`]).
+    pub sub_queries: Vec<DriveSubQuery<'a>>,
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive/src/query/mod.rs:1129]: False positive: Both the entire query module and DriveDocumentQuery already require server or verify, so cbor_query alone never compiles that field.

**Review by QuantumExplorer — COMMENTED** [review]
Approved
